stats/opentelemetry/internal/tracing/custom_carrier_test.go

```diff
@@ -0,0 +1,206 @@
+			setKey:    "key1",
+			setValue:  "newvalue",
+			wantValue: "newvalue",
+			wantKeys:  []string{"key1"},
+		},
+		{
+			name:      "set new key with different existing key",
+			initialMD: metadata.MD{"key2": []string{"value2"}},
+			setKey:    "key1",
+			setValue:  "value1",
+			wantValue: "value1",
+			wantKeys:  []string{"key2", "key1"},
+		},
+		{
+			name:      "set non-grcp-trace-bin binary key",
+			initialMD: metadata.MD{"key1": []string{"value1"}},
+			setKey:    "non-grcp-trace-bin",
+			setValue:  base64.StdEncoding.EncodeToString([]byte{0x01, 0x02, 0x03}),
+			wantValue: "", // non-grcp-trace-bin key should not be set
+			wantKeys:  []string{"key1"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			c := NewCustomCarrier(metadata.NewOutgoingContext(ctx, test.initialMD))
+
+			c.Set(test.setKey, test.setValue)
+
+			gotKeys := c.Keys()
+			equalKeys := cmp.Equal(test.wantKeys, gotKeys, cmpopts.SortSlices(func(a, b string) bool {
+				return a < b
+			}))
+			if !equalKeys {
+				t.Fatalf("got keys %v, want %v", gotKeys, test.wantKeys)
+			}
+			gotMD, _ := metadata.FromOutgoingContext(c.Context())
+			if test.wantValue != "" && gotMD.Get(test.setKey)[0] != test.setValue {
+				t.Fatalf("got value %s, want %s, for key %s", gotMD.Get(test.setKey)[0], test.setValue, test.setKey)
+			}
+			cancel()
+		})
+	}
+}
+
+func (s) TestGetBinary_GRPCTraceBinHeaderKey(t *testing.T) {
+	want := []byte{0x01, 0x02, 0x03}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	c := NewCustomCarrier(stats.SetIncomingTrace(ctx, want))
+
+	got := c.GetBinary()
+	if got == nil {
+		t.Fatalf("GetBinary() = nil, want %v", got)
+	}
+	if string(got) != string(want) {
+		t.Fatalf("GetBinary() = %s, want %s", got, want)
+	}
+}
+
+func TestGet_GRPCTraceBinHeaderKey(t *testing.T) {
+	tc := []byte{0x01, 0x02, 0x03}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	c := NewCustomCarrier(stats.SetIncomingTrace(ctx, tc))
+
+	got := c.Get(GRPCTraceBinHeaderKey)
+	want := base64.StdEncoding.EncodeToString(tc)
+	if got != want {
+		t.Fatalf("Get() = %s, want %s", got, want)
+	}
+}
+
+func (s) TestSetBinary_GRPCTraceBinHeaderKey(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	want := []byte{0x01, 0x02, 0x03}
+	c := NewCustomCarrier(ctx)
+
+	c.SetBinary(want)
+	got := stats.OutgoingTrace(c.Context())
+	if !reflect.DeepEqual(got, want) {
+		t.Fatalf("got %v, want %v", got, want)
+	}
+}
+
+func (s) TestSet_GRPCTraceBinHeaderKey(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	want := []byte{0x01, 0x02, 0x03}
+	c := NewCustomCarrier(ctx)
+
+	c.Set(GRPCTraceBinHeaderKey, base64.StdEncoding.EncodeToString(want))
+	got := stats.OutgoingTrace(c.Context())
+	if !reflect.DeepEqual(got, want) {
+		t.Fatalf("got %v, want %v", got, want)
+	}
+}
```

Comment:
Combine the set and get tests into one t-test (there's ways to allow logic, such as early returns before doing the encoding). And please add more test cases. Test the case where you don't pass in right key. Or get the right key.

Reply:
Mentioned above. Removed these and merge the tests for Get/Set in above t-test. Only TestSetBinary and TestGetBinary are separate as they are different functions